tez-dag/src/main/java/org/apache/tez/dag/api/client/registry/zookeeper/ZkAMRegistry.java

```diff
@@ -0,0 +1,175 @@
+    super("ZkAMRegistry");
+    this.externalId = externalId;
+  }
+
+  @Override
+  public void serviceInit(Configuration conf) {
+    if (zkConfig == null) {
+      zkConfig = new ZkConfig(conf);
+      this.client = zkConfig.createCuratorFramework();
+      this.namespace = zkConfig.getZkNamespace();
+      LOG.info("AMRegistryZkImpl initialized");
+    }
+  }
+
+  @Override
+  public void serviceStart() throws Exception {
+    if (!started) {
+      client.start();
+      started = true;
+      LOG.info("AMRegistryZkImpl started");
+    }
+  }
+
+  //Deletes from Zookeeper AMRecords that were added by this instance
+  @Override
+  public void serviceStop() throws Exception {
+    List<AMRecord> records = new ArrayList<>(amRecords);
+    for (AMRecord amRecord : records) {
+      remove(amRecord);
+    }
+    client.close();
+    LOG.info("AMRegistryZkImpl shutdown");
+  }
+
+  //Serialize AMRecord to ServiceRecord and deliver the JSON bytes to
+  //zkNode at the path:  <TEZ_AM_REGISTRY_NAMESPACE>/<appId>
+  @Override
+  public void add(AMRecord server) throws Exception {
+    RegistryUtils.ServiceRecordMarshal marshal = new RegistryUtils.ServiceRecordMarshal();
+    String json = marshal.toJson(server.toServiceRecord());
+    try {
+      final String path = namespace + "/" + server.getApplicationId().toString();
+      client.setData().forPath(path, json.getBytes(StandardCharsets.UTF_8));
+      LOG.info("Added AMRecord to zkpath {}", path);
+    } catch (KeeperException.NoNodeException nne) {
+      client.create().creatingParentContainersIfNeeded().withMode(CreateMode.EPHEMERAL)
+          .forPath(namespace + "/" + server.getApplicationId().toString(), json.getBytes(StandardCharsets.UTF_8));
+    }
+    amRecords.add(server);
+  }
+
+  @Override
+  public void remove(AMRecord server) throws Exception {
+    amRecords.remove(server);
+    final String path = namespace + "/" + server.getApplicationId().toString();
+    client.delete().forPath(path);
+    LOG.info("Deleted AMRecord from zkpath {}", path);
+  }
+
+  @Override
+  public ApplicationId generateNewId() throws Exception {
+    createNamespaceIfNotExists();
+    long namespaceCreationTime = getNamespaceCreationTime();
+
+    boolean success = false;
+    long startTime = System.currentTimeMillis();
+    RetryPolicy retryPolicy = zkConfig.getRetryPolicy();
+    int tryId = 0;
+    for (int i = 0; (i < zkConfig.getCuratorMaxRetries()) && !success; i++) {
+      List<String> children = client.getChildren().forPath(namespace);
+      if (children != null && !children.isEmpty()) {
+        children.sort(Collections.reverseOrder());
+        String last = children.getFirst();
+        ApplicationId lastAppId = ApplicationId.fromString(last);
+        tryId = lastAppId.getId() + 1;
+      }
+      ApplicationId tryAppId = ApplicationId.newInstance(namespaceCreationTime, tryId);
+      try {
+        client
+            .create()
+            .withMode(CreateMode.EPHEMERAL)
```

Comment:
should it be ``create().withMode(PERSISTENT_SEQUENTIAL)``?

Reply:
EPHEMERAL works here with the current logic (described in a previous comment), auto-deletion is a value